File: sdk/go/ancora/xlang_verifier_test.go
package ancora_test

import (
	"encoding/json"
	"testing"
)

// Cross-language conformance: verifier scenario -- Go.

type xlangVerifierEvent struct {
	Kind        string          `json:"kind"`
	RunID       string          `json:"run_id"`
	ActivityKey string          `json:"activity_key,omitempty"`
	Output      json.RawMessage `json:"output,omitempty"`
}

func makeXlangVerifierEvents(runID string) []xlangVerifierEvent {
	return []xlangVerifierEvent{
		{Kind: "started",   RunID: runID},
		{Kind: "activity",  RunID: runID, ActivityKey: "drafter"},
		{Kind: "activity",  RunID: runID, ActivityKey: "verifier"},
		{Kind: "completed", RunID: runID, Output: json.RawMessage(`{"verdict":"approved"}`)},
	}
}

func TestXlangGoVerifierStartedFirst(t *testing.T) {
	evs := makeXlangVerifierEvents("xlv-go")
	if evs[0].Kind != "started" {
		t.Fatalf("expected started, got %q", evs[0].Kind)
	}
}

func TestXlangGoVerifierCompletedLast(t *testing.T) {
	evs := makeXlangVerifierEvents("xlv-go")
	last := evs[len(evs)-1]
	if last.Kind != "completed" {
		t.Fatalf("expected completed, got %q", last.Kind)
	}
}

func TestXlangGoVerifierDrafterBeforeVerifier(t *testing.T) {
	evs := makeXlangVerifierEvents("xlv-go")
	var keys []string
	for _, e := range evs {
		if e.Kind == "activity" {
			keys = append(keys, e.ActivityKey)
		}
	}
	if len(keys) != 2 || keys[0] != "drafter" || keys[1] != "verifier" {
		t.Fatalf("expected [drafter verifier], got %v", keys)
	}
}

func TestXlangGoVerifierOutputHasVerdict(t *testing.T) {
	evs := makeXlangVerifierEvents("xlv-go")
	last := evs[len(evs)-1]
	var out map[string]interface{}
	if err := json.Unmarshal(last.Output, &out); err != nil {
		t.Fatalf("unmarshal output: %v", err)
	}
	if out["verdict"] != "approved" {
		t.Fatalf("expected verdict=approved, got %v", out["verdict"])
	}
}

File: sdk/ts/__tests__/xlang-verifier.test.ts
// Cross-language conformance: verifier scenario -- TypeScript (offline).

interface XlangVerifierEvent {
  kind: string
  run_id: string
  activity_key?: string
  output?: { verdict: string }
}

const XLANG_VERIFIER_RUN_ID = 'xlv-ts'

function makeXlangVerifierEvents(runId: string): XlangVerifierEvent[] {
  return [
    { kind: 'started',   run_id: runId },
    { kind: 'activity',  run_id: runId, activity_key: 'drafter' },
    { kind: 'activity',  run_id: runId, activity_key: 'verifier' },
    { kind: 'completed', run_id: runId, output: { verdict: 'approved' } },
  ]
}

describe('xlang verifier scenario -- TypeScript', () => {
  const events = makeXlangVerifierEvents(XLANG_VERIFIER_RUN_ID)

  it('started event is first', () => {
    expect(events[0].kind).toBe('started')
  })

  it('completed event is last', () => {
    expect(events[events.length - 1].kind).toBe('completed')
  })

  it('drafter activity comes before verifier', () => {
    const keys = events.filter(e => e.kind === 'activity').map(e => e.activity_key)
    expect(keys).toEqual(['drafter', 'verifier'])
  })

  it('completed output has verdict approved', () => {
    const last = events[events.length - 1]
    expect(last.output?.verdict).toBe('approved')
  })

  it('run_id consistent across events', () => {
    for (const ev of events) {
      expect(ev.run_id).toBe(XLANG_VERIFIER_RUN_ID)
    }
  })
})
